server/src/main/java/com/linecorp/centraldogma/server/command/AbstractCommandExecutor.java

```diff
@@ -97,8 +102,24 @@ public final void setWritable(boolean writable) {
     @Override
     public final <T> CompletableFuture<T> execute(Command<T> command) {
         requireNonNull(command, "command");
-        if (!isWritable()) {
-            throw new IllegalStateException("running in read-only mode");
+        if (!isWritable() && !(command instanceof AdministrativeCommand)) {
+            // Reject all commands except for AdministrativeCommand when the replica is in read-only mode.
+            // AdministrativeCommand is allowed because it is used to change the read-only mode or migrate
+            // metadata under maintenance mode.
+            throw new IllegalStateException("running in read-only mode. command: " + command);
+        }
+
+        if (command.type() == CommandType.UPDATE_SERVER_STATUS) {
+            final UpdateServerStatusCommand command0 = (UpdateServerStatusCommand) command;
+            final boolean writable0 = command0.writable();
+            if (writable0 != writable) {
+                setWritable(writable0);
+                if (writable0) {
+                    logger.warn("Left read-only mode.");
+                } else {
+                    logger.warn("Entered read-only mode.");
+                }
```

Comment:
Question) I'm wondering about the scenario for the usage of this command.
If we issue this command and it is replicated (since there is no `return` statement here), does this mean all replicas will enter read-only mode? Asking because this sounds like a potentially dangerous operation.

Reply:
> If we issue this command and it is replicated (since there is no return statement here), does this mean all replicas will enter read-only mode?

Right. `UpdateServerStatusCommand` is designed to update the whole cluster status.

> Asking because this sounds like a potentially dangerous operation.

Agreed. It is dangerous and carefully used by administrators. The REST API for read-only mode only updates the server's status but does not update other replicas. We had to update each server's status with REST API.

This command would be useful to update the whole cluster's status at once.

Reply:
> I'm wondering about the scenario for the usage of this command.

Sorry, I don't think I still understand the scenario 😅  This is what I am imagining:

- Once the server starts up, the leader will check if migration is complete (I am not sure how this be done though)
- If migration isn't complete, then a `UpdateServerStatusCommand` will be issued automatically and the entire cluster enters readonly mode.
- After `UpdateServerStatusCommand` is issued, `MigrateMirrorCommand` is issued
  - `MigrateMirrorCommand` will loop over all projects and migrate each project if not migrated yet
- After the `MigrateMirrorCommand`, another `UpdateServerStatusCommand` will be issued

Is my understanding correct? Or were you imagining that each command be issued manually be administrators?